packages/autorest.python/autorest/codegen/templates/model_base.py.jinja2

```diff
@@ -499,21 +520,53 @@ class Model(_MyMutableMapping):
                 base.__mapping__[discriminator or cls.__name__] = cls  # type: ignore  # pylint: disable=no-member
 
     @classmethod
-    def _get_discriminator(cls) -> typing.Optional[str]:
+    def _get_discriminator(cls, exist_discriminators) -> typing.Optional[str]:
         for v in cls.__dict__.values():
-            if isinstance(v, _RestField) and v._is_discriminator:  # pylint: disable=protected-access
+            if isinstance(v, _RestField) and v._is_discriminator and v._rest_name not in exist_discriminators:  # pylint: disable=protected-access
                 return v._rest_name  # pylint: disable=protected-access
         return None
 
     @classmethod
-    def _deserialize(cls, data):
+    def _deserialize(cls, data, exist_discriminators):
         if not hasattr(cls, "__mapping__"):  # pylint: disable=no-member
             return cls(data)
-        discriminator = cls._get_discriminator()
-        mapped_cls = cls.__mapping__.get(data.get(discriminator), cls)  # pylint: disable=no-member
+        discriminator = cls._get_discriminator(exist_discriminators)
+        exist_discriminators.append(discriminator)
+        mapped_cls = cls.__mapping__.get(
+            data.get(discriminator), cls
+        )  # pylint: disable=no-member
         if mapped_cls == cls:
             return cls(data)
-        return mapped_cls._deserialize(data)  # pylint: disable=protected-access
+        return mapped_cls._deserialize(data, exist_discriminators)  # pylint: disable=protected-access
+
+    def as_dict(self, *, exclude_readonly: bool = False, exclude_none: bool = False) -> typing.Dict[str, typing.Any]:
+        result = {}
+        if exclude_readonly:
+            readonly_props = [p._rest_name for p in self._attr_to_rest_field.values() if _is_readonly(p)]
+        for k, v in self.items():
+            if exclude_readonly and k in readonly_props:
+                continue
+            if exclude_none and (v is None or isinstance(v, _Null)):
+                continue
+            result[k] = Model._as_dict_value(v, exclude_readonly=exclude_readonly, exclude_none=exclude_none)
+        return result
+
+    @staticmethod
+    def _as_dict_value(v: typing.Any, exclude_readonly: bool = False, exclude_none: bool = False) -> typing.Any:
```

Comment:
is there anyway that we can combine this code with the serialization logic we currently have? Basically it seems we have two portions of code that both try to serialize to a dict: 1. is the existing infrastructure to have a model exist as a dict as well and 2. this section which is serializing to dict. Does calling `dict(self)` not work?

Reply:
In order to support mix using the dict and the model properties, we need to store the model in the inner map. For example:
```python
class ModelA(Model):
    prop1: List[ModelB] = rest_field()
    prop2: ModelC = rest_field()
class ModelB(Model):
    prop3: str = rest_field()
class ModelC(Model):
    prop4: str = rest_field()

test = ModelA({"prop1": [{"prop3":"test3"}], "prop2": {"prop4": "test4"}}
test["prop1"][0].prop3 = "test3"  # we need to store prop1's element as model to keep model info
test["prop2"].prop4 = "test4"  # we need to store prop2 as model as well
```